### ISSUE TITLE
Speed-up docker builds in CI

### ISSUE THREAD

**Issue by generall** [qdrant/qdrant#1812]
**Is your feature request related to a problem? Please describe.**
Currently, the slowest part of CI checks is docker builds - it takes about 30 minutes to complete.
That sometimes slows the process of development, cause we need to wait long time even if only a small feature changed.

**Describe the solution you'd like**
Try to apply some of the known techniques to speedup the docker builds:

- caching
- disable `lto = "fat"` for CI builds (not for production)
- ???

**Describe alternatives you've considered**
Tried to use `mold` linked in CI, it didn't really help - https://github.com/qdrant/qdrant/pull/1811
Maybe the setting was wrong, cause locally it helps a lot.

**Additional context**

It would be ideal if docker builds would take less or equal time to other CI tasks, like integration tests.
Expected minimal improvement is about 30min -> 20min of docker CI tasks

**Comment by harshbisle** [comment]
Hey @generall can I give this a shot?

**Comment by generall** [comment]
> Hey @generall can I give this a shot?

sure, feel free, thanks!

**Comment by harshbisle** [comment]
Thanks a lot, on it :smiley:

**Comment by harshbisle** [comment]
While I agree with your point, why don't we go for `lto = "thin"` instead of removing it?

**Comment by generall** [comment]
it is also fine, i think. The question, is how to switch it in CI docker builds only

**Comment by harshbisle** [comment]
By using environment variables, whether to use `lto = "fat"` or `lto = thin`?

**Comment by generall** [comment]
maybe, let's try

**Comment by MrBeeMovie** [comment]
@generall 
Just a question regarding mold linker. Did you try putting
```
[target.x86_64-unknown-linux-gnu]
rustflags = ["-C", "link-arg=-fuse-ld=/path/to/mold"]
```
for each target in cargo.toml instead of trying to run cargo build through mold -run? Mold -run is a fallback if there is no easy way to specify an alternative linker which is not necessary since rust provides an easy way to do so.

**Comment by generall** [comment]
i did not

**Comment by MrBeeMovie** [comment]
> i did not

So just as an example you could do something like [this](https://github.com/MrBeeMovie/qdrant/commit/59f4164ad1a3059122f03493447917b51717f873).

**Comment by JojiiOfficial** [comment]
What about not doing release builds for testing? Seems like it works in debug as well and I get somewhat around 24-25 minutes (without LTO and some other tweaks): https://github.com/JojiiOfficial/qdrant/actions/runs/4863555648

**Comment by generall** [comment]
25 minutes is still too long

**Comment by MrBeeMovie** [comment]
You could also try running build with the [--timings flag](https://doc.rust-lang.org/cargo/reference/timings.html) to get an idea where time is spent during compilation. I will run on local and see if there are any glaring issues I can find.

**Comment by MrBeeMovie** [comment]
> My two cents, I'm not much familiar with Rust so ignore if it is wrong. I think the docker layer caching is not working. I see you have used cargo-chef to cache. GitHub actions run in a completely new container where the previously cached layer is not present. Therefore, cargo-chef is not able to do its work. Something like this may help. https://docs.docker.com/build/cache/backends/gha/

Yeah, this seems to be the case look at this [issue ](https://github.com/rust-lang/cargo/issues/2644#issuecomment-1500464444) in particular.

**Comment by zach-hill** [comment]
> My two cents, I'm not much familiar with Rust so ignore if it is wrong. I think the docker layer caching is not working. I see you have used cargo-chef to cache. GitHub actions run in a completely new container where the previously cached layer is not present. Therefore, cargo-chef is not able to do its work. Something like this may help. [docs.docker.com/build/cache/backends/gha](https://docs.docker.com/build/cache/backends/gha/)

Cargo chef can be used, the github action in this case is just forwarding the inputs "cacheTo" and "cacheFrom" as ("cache-to" and "cache-from" respectively) args to the buildx command. If you're running the buildx command explicitly in your own action, then you would just make sure to use the appropriate arguments(and the driver type if needed for the type of caching).

**Comment by goller** [comment]
Howdy, I used [depot.dev](https://depot.dev/) and was able to get the un-cached build down to around [15 minutes](https://github.com/goller/qdrant/actions/runs/4894483607/jobs/8738768946) with no Dockerfile changes.
The depot client is a drop-in replacement for the docker client, so no scripts need to change.

I work at depot trying to make builds faster.
Depot provides a hosted buildkit service just like the one used in docker but on native machines with persistent disks.

At my previous start-up we built a rust database with really long build times and spent a lot of time trying to make it fast.  The build time problem we had seems really familiar.

If you are interested in this I'd be happy to open up a PR.

**Comment by JojiiOfficial** [comment]
@goller I think its about the integration tests which take 30 minutes in your fork too.

**Comment by goller** [comment]
> @goller I think its about the integration tests which take 30 minutes in your fork too.

Ah, I'll turn on depot for the integration tests too and see what timings we get.
